File: python/BoatsToSavePeople881.py
# You are given an array people where people[i] is the weight of the ith person, and an infinite number of boats
# where each boat can carry a maximum weight of limit. Each boat carries at most two people at the same time,
# provided the sum of the weight of those people is at most limit. Return the minimum number of boats to carry every
# given person.
from typing import List


def numRescueBoats(people: List[int], limit: int) -> int:
    people.sort()
    result, left, right = 0, 0, len(people)
    while left < right:
        result += 1
        right -= 1
        if limit - people[right] >= people[left]:
            left += 1
    return result


if __name__ == '__main__':
    print(numRescueBoats([1, 2], 3))  # 1
    print(numRescueBoats([3, 2, 2, 1], 3))  # 3
    print(numRescueBoats([3, 5, 3, 4], 5))  # 4
    print(numRescueBoats([3, 5, 3, 4], 6))  # 3
    print(numRescueBoats([3, 5, 3, 4], 7))  # 3
    print(numRescueBoats([3, 5, 3, 4], 8))  # 2
    print(numRescueBoats([3, 5, 3, 4], 9))  # 2
    print(numRescueBoats([3, 5, 3, 4], 10))  # 2
    print(numRescueBoats([3, 5, 3, 4], 11))  # 2
    print(numRescueBoats([3, 5, 3, 4], 12))  # 2
    print(numRescueBoats([3, 5, 3, 4], 13))  # 2
    print(numRescueBoats([3, 5, 3, 4], 14))  # 2
    print(numRescueBoats([3, 5, 3, 4], 15))  # 2
    print(numRescueBoats([3, 5, 3, 4], 16))  # 2
    print(numRescueBoats([3, 5, 3, 4], 17))  # 2
    print(numRescueBoats([3, 5, 3, 4], 18))  # 2
    print(numRescueBoats([3, 5, 3, 4], 19))  # 2
    print(numRescueBoats([3, 5, 3, 4], 20))  # 2
    print(numRescueBoats([3, 5, 3, 4], 21))  # 2

File: java/BoatsToSavePeople881.java
import java.util.Arrays;

/*
You are given an array people where people[i] is the weight of the ith person, and an infinite number of boats where each boat can carry a maximum weight of limit. Each boat carries at most two people at the same time, provided the sum of the weight of those people is at most limit.
Return the minimum number of boats to carry every given person.
*/
public class BoatsToSavePeople881 {
    public static void main(String[] args) {
        int[] people = {3, 2, 2, 1};
        int limit = 3;
        System.out.println(numRescueBoats(people, limit)); // 3

        people = new int[]{3, 5, 3, 4};
        limit = 5;
        System.out.println(numRescueBoats(people, limit)); // 4
    }

    public static int numRescueBoats(int[] people, int limit) {
        Arrays.sort(people);
        int result = 0, left = 0, right = people.length;
        while (left < right) {
            result++;
            if (limit - people[--right] >= people[left]) left++;
        }
        return result;

//        Arrays.sort(people);
//        int result = 1, left = 0, right = people.length - 1, currentLimit = limit, currentPeople = 0;
//        while (left <= right) {
//            if (currentLimit >= people[right] && currentPeople < 2) {
//                currentLimit -= people[right--];
//                currentPeople++;
//            } else if (currentLimit >= people[left] && currentPeople < 2) {
//                currentLimit -= people[left++];
//                currentPeople++;
//            } else {
//                currentLimit = limit;
//                currentPeople = 0;
//                result++;
//            }
//        }
//        return result;
    }
}
